### DISCUSSION TITLE
Automatically detecting ports

### DISCUSSION THREAD

**Discussion by mauricioszabo** [mauricioszabo/atom-chlorine#238] [Ideas]
So, this PR (https://github.com/mauricioszabo/atom-chlorine/pull/235) by @rafd added a way to auto-fill the port for Socket REPL, nREPL, and Shadow-CLJS by default.

But I saw some issues while using it:
1. Chlorine (and Clover also) is a package that wants you to connect to Socket REPL
2. There's no specific config, or anything, that creates a `.socket-repl.port` on the root file. In fact, there's nothing on the specification of Socket REPLs at all to create this file
3. But there *is* a specification for nREPL to create this port

This means that, in practice, Chlorine will default to nREPL - that's exactly the protocol he's not great at handling. I'm really thinking on alternatives, maybe even hide this "suggest port" over a config file, because keeping it the way it is, it's not painting a good image on the plug-ins' capabilities IMHO...

**Comment by seancorfield** [comment]
Perhaps add a setting to select the dot file to look for? I suspect a given user is going to use a specific type of REPL for nearly all of their projects anyway. I'm never going to want Chlorine/Clover to use `.nrepl-port`; I'm only ever going to want it to use `.socket-repl-port`.

So perhaps a text field that folks can put a comma-separated list of dot-file names in order to search for?

**Reply by seancorfield** [reply]
Although `.nrepl-port` should only be present if an nREPL is actually running, right? And Chlorine searches for the Shadow socket file, then (my) `.socket-repl-port` file, then `.nrepl-port` so, for me, it'll never find `.nrepl-port` -- that's never in any of my folders, but `.socket-repl-port` is inside nearly all of them.

**Reply by mauricioszabo** [reply]
Yes, but the problem is that `lein repl` will always create the `.nrepl-port`. So, for people that don't create the `.socket-repl-port` manually (or via a script) and use `lein`, the `.nrepl-port` will always be the first options.

I'm really thinking about adding a configuration parameter so that the `.nrepl-port` is never used, and if someone wants to use it, he can just enable the config...

**Reply by rafd** [reply]
What's wrong with using nREPL over Socket REPL for lein users?

**Reply by mauricioszabo** [reply]
@rafd nothing wrong per-se, but Chlorine shines on Socket-REPL (and there are lots of experiments to add more sweet things on Socket-REPL).

nREPL support will probably never be _that good_ on Chlorine mostly because it's not the package's focus. So, if a newcomer sees Chlorine, starts the Socket REPL, and tries to connect to the REPL, he'll see the port auto-filled for him and think "well, that's the port I'm going to connect" and then be presented with an worse experience that he have on Calva, for example. That's what's worrying me to be honest.

In fact, nREPL support only exist because it was confusing for people to connect to the port that Lein gave then, and nothing happening on Atom....

**Reply by seancorfield** [reply]
Yeah, I think at this point you've convinced me it really does need to be an opt-in setting. Sounds like it could work automatically for Shadow, and it could also work automatically for `.socket-repl-port` if that happens to be present, but `.nrepl-port` would need to be strictly opt-in? So maybe a text-based setting field with filenames to check for in order but with the default being empty?

**Reply by mauricioszabo** [reply]
Yes, I believe the order should be `.socket-repl-port`, then `.shadow-cljs`, and then, if there's a config, `.nrepl-port`.

I think I'll add a config "Auto-detect nREPL port". In this case, it'll skip `.nrepl-port` completely unless the config is activated. I'll migrate the code to REPL-Tooling and issue a new version of Chlorine and Clover with this new code.

**Reply by rafd** [reply]
Some other options to consider:
 - recommend that people manually create a `.socket-repl-port` file (since you're already asking them to add the relevant config to `deps.edn`)
 - show a checkbox list of options when multiple sources detected, always including Socket REPL (empty by default)
 - try to parse `deps.edn` for the port

**Reply by rafd** [reply]
Out of curiosity, what additional advantage does Socket REPL currently have over nREPL?

**Reply by seancorfield** [reply]
@rafd I think you're starting to over-complicate things at this point: Chlorine/Clover are great because they are _simple_ and generally just "do the right thing" without a lot of options and interactions.

Re: Socket REPL -- the great advantage for me is that socket REPL is built into Clojure itself and requires no code and no dependencies: I can use just a JVM option when starting any process to get Clojure to start a socket REPL on a given port (or even have it auto-select an available port). I can then use Chlorine/Clover to connect to _any_ Clojure process that is running a socket REPL -- a local one for dev, a remote one (such as on production) for debugging -- and it's the exact same editor experience. No extra libraries, no dependencies.

**Reply by mauricioszabo** [reply]
@rafd to be honest, nREPL is better as a tooling assist. But it only works well for Babashka and Clojure, and even Clojure needs an additional dependency (something that may be not present on production environments for example, if you want to remotely debug code).

On Clojerl, Clojure.CLR, Plank, Lumo, Joker, and experimental Clojure implementations they always add a Socket REPL, so Chlorine *can be used* to connect to "esoteric Clojure implementations". Apart from that, nREPL is better in any other way for tooling.

Chlorine (and Clover) really shine because they use some interesting techniques over Socket REPL. Yes, it's harder to program on it, and I would probably be better by using nREPL, but there are already *lots of packages* for nREPL and I don't want to reinvent the wheel or take their places.

So, the _focus_ of this package is Socket REPL. It works with nREPL also, but it's more like a "nice-to-have" and it still have its quirks, so that's the reason I'm afraid to make it an "easier option" for people to use :)

**Reply by mauricioszabo** [reply]
Also, nREPL over CojureScript is kind of a pain, really...
